### PR TITLE
Bug 1752045: UPSTREAM: <carry>:Add a RBAC checker for external IP ranger

### PR DESCRIPTION
**Pull request by abhat**
This PR adds a RBAC checker for adding an External IP range for a service.

### CODE REVIEW

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go [hunk 1]
```diff
@@ -167,5 +183,29 @@ func (r *externalIPRanger) Validate(a admission.Attributes, _ admission.ObjectIn
...
+
 	return nil
 }
+
+func (r *externalIPRanger) checkAccess(attr admission.Attributes) (bool, error) {
+	authzAttr := authorizer.AttributesRecord{
+		User:            attr.GetUserInfo(),
+		Verb:            "create",
+		Namespace:       attr.GetNamespace(),
+		Resource:        "endpoints",
```
Comment (danwinship) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: "service"

**Comment by soltysh** [comment]
/cc @deads2k @sttts 
since this is your area and on top of this is adding another carry patch

**Comment by squeed** [comment]
/retitle Bug 1752045: UPSTREAM: <carry>:Add a RBAC checker for external IP ranger

**Comment by squeed** [comment]
FWIW, this used to live in pkg/network, but since it's part of the kube-apiserver, vendor/... is the authoritative source.

**Comment by abhat** [comment]
Update: fixed a typo in the testname in one of the admission tests in the recent force-push.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go [hunk 2]
```diff
@@ -155,6 +171,7 @@ func (r *externalIPRanger) Validate(a admission.Attributes, _ admission.ObjectIn
 			ip := net.ParseIP(s)
 			if ip == nil {
 				errs = append(errs, field.Forbidden(field.NewPath("spec", "externalIPs").Index(i), "externalIPs must be a valid address"))
+				invalidIp = true
```
Comment (danwinship) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: Kubernetes [already validates](https://github.com/openshift/origin/blob/17f1107c/vendor/k8s.io/kubernetes/pkg/apis/core/validation/validation.go#L3712) that all `service.Spec.ExternalIPs` values are valid IPs, so we don't need to check that here. So we don't need to distinguish admin-overridable from non-admin-overridable errors.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go [hunk 3]
```diff
@@ -164,8 +181,37 @@ func (r *externalIPRanger) Validate(a admission.Attributes, _ admission.ObjectIn
...
-		return apierrs.NewInvalid(a.GetKind().GroupKind(), a.GetName(), errs)
+		//if there are errors reported, resort to RBAC check to see
+		//if this is an admin user who can over-ride the check
+		if !invalidIp {
+			allow, err := r.checkAccess(a)
+			if err != nil {
+				return err
+			}
+			if !allow {
+				return apierrs.NewInvalid(a.GetKind().GroupKind(), a.GetName(), errs)
```
Comment (danwinship) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: as in the restricted endpoints controller, this should be `apierrs.NewForbidden`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission_test.go [hunk 4]
```diff
@@ -1,15 +1,45 @@
...
+}
+
+func (a *fakeTestAuthorizer) Authorize(attributes authorizer.Attributes) (authorizer.Decision, string, error) {
+	ui := attributes.GetUser()
+	if ui == nil {
+		return authorizer.DecisionNoOpinion, "", fmt.Errorf("No valid UserInfo for Context")
+	}
+	// system:serviceaccount:herpy:derpy user aka admin user is allowed to set
+	// external IPs
+	if ui.GetName() == "system:serviceaccount:herpy:derpy" {
```
Comment (danwinship) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission_test.go]: use a more self-documenting name. No one is going to understand why a test is expected to pass for "herpy:derpy" but fail for "slurpy:derpy".

eg `"system:serviceaccount:test:admin"` vs `"system:serviceaccount:test:ordinary-user"`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go (hunk 3 above)
Comment (abhat) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: I only moved the existing error inside the `!allow` block. But for correctness, I can change it to `apierrs.NewForbidden`

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission_test.go (hunk 4 above)
Comment (abhat) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission_test.go]: I was following the example in the `podnodeconstraints` admission plugin, but I suppose I can change the account name for clarity. The fake authorizer comments do make it clear that `herpy:derpy` is an admin user, but i guess i can make it unambiguous further using appropriate account names. Thanks for keeping me honest.

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go (hunk 2 above)
Comment (abhat) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: Isn't it better to validate them early in the cycle before it gets to the spec level validation?

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go (hunk 3 above)
Comment (abhat) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: Actually, the only apierror that takes an error list as argument is `NewInvalid`. Do we want to use `NewForbidden` and only pick the first element in the `errs` slice?

**Comment by deads2k** [comment]
/hold

I'd like to see an enhancement for this that outlines the problem, links prior art, and suggests different options for the RBAC permissions.

**Comment by deads2k** [comment]
1 Is this really a namespace scoped check?
2. Does subdivision of this power by namespace make sense?
3. Should regular users with this power in one namespace be able delegate that power to other users?
4. Is this an escalating permission for the sake of authorization scope grants?

**Comment by danwinship** [comment]
We don't really expect anyone to actually be granted permission to do this. The RBAC check is just an escape hatch for cluster admins to be able to bypass the checks if absolutely needed. The design was based on what the restricted endpoints admission controller does, but that's a few years old and maybe that's no longer considered the right thing?

That said:

> 1 Is this really a namespace scoped check?
> 2. Does subdivision of this power by namespace make sense?
> 3. Should regular users with this power in one namespace be able delegate that power to other users?

No to all three: being able to bypass external IP restrictions in any namespace gives you the power to attack people in every namespace.

> 4. Is this an escalating permission for the sake of authorization scope grants?

I don't know what you mean by that

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go (hunk 2 above)
Comment (danwinship) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: validation should happen before admission controllers; there's no point in doing admission control on an object that will be rejected anyway

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go (hunk 3 above)
Comment (danwinship) [reply vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: It seems like it should have been Forbidden before... anyway, you could aggregate the errors into a single message perhaps

File: vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go [hunk 5]
```diff
@@ -165,8 +178,31 @@ func (r *externalIPRanger) Validate(ctx context.Context, a admission.Attributes,
...
+		if !allow {
+			return admission.NewForbidden(a, errs.ToAggregate())
+		}
 	}
+
 	return nil
 }
+
+func (r *externalIPRanger) checkAccess(attr admission.Attributes) (bool, error) {
+	authzAttr := authorizer.AttributesRecord{
```
Comment (deads2k) [inline vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/network/externalipranger/externalip_admission.go]: this matches our design.

**Pull request merged**